Create a standalone TimeLineProvider and TimelineEntry for the lock screen

tzef opened the pull request:
This is the PR for the refactoring of the timeline provider for lock screen widgets.

1.  https://github.com/wordpress-mobile/WordPress-iOS/pull/20309 (✅ Approved)
2. https://github.com/wordpress-mobile/WordPress-iOS/pull/20312 (✅ Approved) 
3. https://github.com/wordpress-mobile/WordPress-iOS/pull/20342 (TBD) 
4. https://github.com/wordpress-mobile/WordPress-iOS/pull/20353 (✅ Approved)
5. https://github.com/wordpress-mobile/WordPress-iOS/pull/20371 (✅ Approved)
6. https://github.com/wordpress-mobile/WordPress-iOS/pull/20317 (✅ Approved)
7. https://github.com/wordpress-mobile/WordPress-iOS/pull/20368 (✅ Approved) 👈 you're here!
8. https://github.com/wordpress-mobile/WordPress-iOS/pull/20399 (✅ Approved)
9. https://github.com/wordpress-mobile/WordPress-iOS/pull/20405 (✅ Approved)
(Confirm the data display on UI correctly in this phase)
10. https://github.com/wordpress-mobile/WordPress-iOS/pull/20422 (✅ Approved)
11. https://github.com/wordpress-mobile/WordPress-iOS/pull/20427 (In Reviewing)

## Description
While it is working to reuse the home screen timeline provider and entry for the lock screen, there are some differences in logic. For instance, the disable check and status are redundant for the lock screen widget.

Creating a timeline provider and entry specific to the lock screen to separate the concerns, which should be helpful for improving understandability and maintainability. Here are the steps I took to achieve this:

1. Add `LockScreenStatsWidgetData` to define the properties the lock screen widgets need.
2. Add `LockScreenSiteListProvider` to handle the `LockScreenStatsWidgetEntry` timeline.
3. Extract the widget load to `WidgetDataReader` for reusability and testability.
4. Expand the result type in `StatsWidgetsService` for sharing the fetchStats with two providers.
5. Update the view provider and view to be based on the lock screen widget, removing unnecessary type handling.

## Testing Instructions
This refractor did not introduce any new features. The following tests are for regression testing purposes.
### In WordPress target
#### Won't impact existing widgets
1. Added today, all-time, and this-week home screen widgets
2. Upgrade the app
3. Existing widgets still work well
#### New widget can be added
1. Upgrade the app
2. WordPress in the widget selection page in the home screen editor
3. Able to add today, all-time, and this-week widgets and display the correct stats data
#### Status switching
- Display disable message when fwf off
- Display no-site message when login with a new account without setup websites
- Display logged out message when logout the account

### In Jetpack target
#### Won't impact home screen widgets 
- As same as WordPress testing instructions
#### Won't impact lock screen widgets (refresh mechanism not implemented yet)
- Existing lock screen widgets work well after upgrading the app
- Able to add today views widget to the lock screen and display the correct stats data
#### Status switching
- Jetpack others status not implemented yet, status switching testing would in next PR

## Regression Notes
1. Potential unintended areas of impact
Home screen widgets

5. What I did to test those areas of impact (or what existing automated tests I relied on)
Refer to the above testing Instructions

6. What automated tests I added (or what prevented me from doing so)
`WidgetDataReaderTests` for testing status check

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/LockScreenSiteListProvider.swift:
```diff
@@ -0,0 +1,93 @@
+        completion(.siteSelected(widgetData, context))
+    }
+
+    func getTimeline(for configuration: SelectSiteIntent, in context: Context, completion: @escaping (Timeline<LockScreenStatsWidgetEntry>) -> Void) {
```
There're many little details in this logic crafted over time. Both logic checks as well as caching functionality. 🤔

Do you see any potential risks to having this (almost) duplicated? If yes, do those risks merit having a separate provider?

tzef replied:
Yes, it's absolutely a bad smell. Many details are highly duplicated I just copied from another provider. It's a high risk to cause inconsistent behavior and not easy to read because it's mixed with different level logic.

In my opinion, the timeline provider is the component closest to the use case, but some details, such as the cache, are obviously not at the same level. I assume these low-level logic components should rarely change when introducing a new timeline provider for the new widgets.

Extracting them to a standalone data reader could protect them from easy changes and make them testable due to no dependencies on WidgetKit and SwiftUI.

staskus commented on WordPress/WordPressStatsWidgets/Helpers/WidgetDataReader.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+final class WidgetDataReader<T: HomeWidgetData> {
```
I appreciate what is done with this reader, reusing common functionality for the providers. Do you see any more opportunities for such reuse? Would they make sense in your mind?

tzef replied:
Yes, theoretically all duplicated code could be extracted for reuse using a similar approach. I was thinking about what the reasonable responsibility of the timeline provider is.

As a final decision, I decided to leave the refresh time and legacy data check to fetch the API in the timeline provider, focusing on the refresh frequency and the API data source, does it make sense to you?

staskus replied:
> As a final decision, I decided to leave the refresh time and legacy data check to fetch the API in the timeline provider, focusing on the refresh frequency and the API data source, does it make sense to you?

I looked through the code, I think it's a reasonable balance 👍

staskus reviewed (commented):
I know this is still a draft, posting interesting questions.

Let me know when it's ready for review. Thanks!

tzef committed:
Add widget data and timeline entry for lock screen to separate concerns

tzef committed:
Add lock screen time line provider for lock screen stats widget entry

Update stats widgets service result type to consist of home widget and lock screen widget for reusable

tzef committed:
Update to use LockScreenSiteListProvider in LockScreenStatsWidget

Update associated type widget data to conform `LockScreenStatsWidgetData` more

Update model mapper and view provider to based on lock screen widget data

Remove unnecessary `disabled` status from `LockScreenStatsWidgetsView`

Remove unnecessary `widgetKind` parameter from `LockScreenStatsWidget`

Remove unused statsURL converter from the view model mapper

Add home widget data conform lock screen widget data extension to WordPress target for unitTest

tzef committed:
Extract widget load data to loader for sharing with lock screen time line provider

tzef committed:
Update widget reader to extract the cache read from getSnapshot

tzef committed:
Extract noData, noSite, loggedOut, disabled status checking logic to data reader with status callback

tzef committed:
Add unitTest for widget data reader

Extract cache reading to `HomeWidgetDataFileReader` for mocking the testing object

Extract app configuration is jetpack to parameter for mocking the config

tzef committed:
Remove multiple space lines

tzef commented on WordPress/WordPressStatsWidgets/Helpers/WidgetDataReader.swift:
```diff
@@ -0,0 +1,75 @@
+    ///   - configuration: Configuration of the Widget Site Selection Intent
+    ///   - defaultSiteID: ID of the default site in the account
+    /// - Returns: Widget data
+    func widgetData(for configuration: SelectSiteIntent, defaultSiteID: Int?) -> T? {
```
Updated this function parameter `defaultSiteID` optional to make provider can reuse the same function in `getSnapshot` to avoid using the cache reader directly.

And I think `if else` condition is more readable than the double `??` as existing
```
widgetData(for: configuration.site?.identifier ?? String(defaultSiteID)) ?? widgetData(for: String(defaultSiteID))
```

staskus replied:
> to make provider can reuse the same function in getSnapshot to avoid using the cache reader directly.

Good catch to use it for both `getSnapshot` and `getTimeline` 👍

> And I think if else condition is more readable than the double ?? as existing

I agree. Thanks for expanding it!

tzef commented on WordPress/WordPressStatsWidgets/Helpers/HomeWidgetDataFileReader.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+final class HomeWidgetDataFileReader: WidgetDataCacheReader {
```
Extracting the logic for testing the widget data reader

tzef reviewed (commented):
@staskus , as this PR changed existing provider quite much, I will do more testings, and you can take the review in the meanwhile, thanks 🙏

staskus commented on WordPress/WordPressStatsWidgets/Extensions/HomeWidgetData+LockScreenStatsWidgetData.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+extension HomeWidgetTodayData: LockScreenStatsWidgetData {
```
👍

staskus commented on WordPress/WordPressStatsWidgets/Helpers/WidgetDataReader.swift:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    func widgetData(
```
The logic looks to have remained the same. 

I think that's a good extraction of the code 👍

staskus commented on WordPress/WordPressStatsWidgets/Helpers/WidgetDataReader.swift:
```diff
@@ -0,0 +1,75 @@
+        for configuration: SelectSiteIntent,
+        defaultSiteID: Int?,
+        isJetpack: Bool,
+        onDisabled: (() -> Void)? = nil,
```
Instead of many callbacks, what would you think about the `Result` type? Or a custom `enum` with many cases? I think it would correctly indicate 2 things:

- The method is synchronous 
- The method only has one return type at the time

Maybe it would also help with testability? Do you think this is a change worth making?

tzef replied:
Yup, it makes sense, I also feel multiple callbacks can't guarantee one return, so that's why I write the unit test,
I agree with `Result` is better, I will submit an update tomorrow 👍

staskus replied:
Okay. Thanks!

staskus commented on WordPress/WordPressStatsWidgets/Helpers/WidgetDataReader.swift:
```diff
@@ -0,0 +1,75 @@
+            return
+        }
+
+        onSiteSelected(widgetData)
```
I would offer to rename this callback / case. From the outside, it reads as if someone has made a site selection. It's more of a `success` case that returns `widgetData`.

tzef replied:
I see, I will try to improve it, thanks for the opinions

staskus commented on WordPress/WordPressStatsWidgets/SiteListProvider.swift:
```diff
@@ -17,109 +17,57 @@ struct SiteListProvider<T: HomeWidgetData>: IntentTimelineProvider {
-        }
-
-        completion(.siteSelected(widgetData, context))
+        let content = widgetDataLoader.widgetData(for: configuration, defaultSiteID: defaultSiteID) ?? placeholderContent
```
I'm happy that this change will make `SiteListProvider` more readable as well!

staskus commented:
@tzef thanks for the work! 👍

I reviewed the extracted timeline code and submitted my comments so you could check them as early as possible. 

I'll test the solution and also check the tests you've written.

staskus commented:
Tested.

✅ WordPress/Jetpack home widget regression tests passed. All the different states appear correctly.
✅ New widgets show the data. As you mentioned, data refresh is still missing.

tzef committed:
Update multiple callback to Result and enum Error

tzef commented:
Updated the code to use `Result` and a custom `enum` for exception status. This makes the caller side much more straightforward than using callbacks. Thank you for the great suggestions. 👍

staskus reviewed (approved):
@tzef thanks for the changes! Looks good. The `success` case now is very explicit and clear.

I'm rerunning the CI and merging if all is good.

staskus commented:
@tzef merged
